src/Timer/Timer.c

```diff
@@ -0,0 +1,381 @@
+    *timerReloadRegister = config->reloadValue;
+}
+
+void
+Timer_getConfig(const uint32_t timerControlRegister, const uint32_t timerReloadRegister, Timer_Config *const config)
+{
+    config->isEnabled = Timer_getFlag (timerControlRegister, TIMER_CONTROL_EN);
+    config->isInterruptEnabled = Timer_getFlag (timerControlRegister, TIMER_CONTROL_IE);
+    config->isAutoReloaded = Timer_getFlag (timerControlRegister, TIMER_CONTROL_RS);
+    config->isChained = Timer_getFlag (timerControlRegister, TIMER_CONTROL_CH);
+    config->reloadValue = timerReloadRegister;
+}
+
+void
+Timer_start(volatile uint32_t *const timerControlRegister)
+{
+    Timer_setFlag (timerControlRegister, TIMER_FLAG_SET, TIMER_CONTROL_LD);
+    Timer_setFlag (timerControlRegister, TIMER_FLAG_SET, TIMER_CONTROL_EN);
+}
+
+void
+Timer_stop(volatile uint32_t *const timerControlRegister)
+{
+    Timer_setFlag (timerControlRegister, TIMER_FLAG_RESET, TIMER_CONTROL_EN);
+}
+
+bool
+Timer_hasFinished(const uint32_t timerControlRegister, const uint32_t timerCounterRegister) {
+    bool result = false;
+    if (!Timer_getFlag (timerControlRegister, TIMER_CONTROL_EN) && timerCounterRegister == TIMER_UNDERFLOWED) {
+        result = true;
+    }
+    
+    return result;
+}
+
+void
+irqInit(rtems_interrupt_entry *entry, Timer_InterruptHandler *handler, const uint8_t irqNumber)
+{
+    rtems_interrupt_clear(irqNumber);
+    rtems_interrupt_entry_initialize(
+            entry,
+            (rtems_interrupt_handler)Timer_handleIrq,
+            handler,
+            "Timer Interrupt");
+    rtems_interrupt_entry_install(
+        irqNumber,
+        RTEMS_INTERRUPT_UNIQUE,
+        entry);
+    rtems_interrupt_vector_enable(irqNumber);
+}
+
+void
+irqDeinit(rtems_interrupt_entry *entry, const uint8_t irqNumber)
+{
+    rtems_interrupt_vector_disable(irqNumber);
+    rtems_interrupt_entry_remove(irqNumber, &entry);
+}
+
+void Timer_handleIrq(Timer_InterruptHandler* const handler)
+{
+    handler->callback(handler->arg);
+}
+
+static inline Timer_Apbctrl1_Registers
+getApbctrl1TimerAddressById(Timer_Id id)
+{
+    Timer_Apbctrl1_Registers result = NULL;
+
+    switch (id) {
+        case Timer_Id_1:
+            result = (Timer_Apbctrl1_Registers) Timer1_Apbctrl1_Address;
+            break;
+        case Timer_Id_2:
+            result = (Timer_Apbctrl1_Registers) Timer2_Apbctrl1_Address;
+            break;
+        case Timer_Id_3:
+            result = (Timer_Apbctrl1_Registers) Timer3_Apbctrl1_Address;
+            break;
+        case Timer_Id_4:
+            result = (Timer_Apbctrl1_Registers) Timer4_Apbctrl1_Address;
+            break;
+        default:
+            result = (Timer_Apbctrl1_Registers) TimerInvalid_Apbctrl1_Address;
+    }
+
+    return result;
+}
+
+void
+Timer_Apbctrl1_init(Timer_Id id, Timer_Apbctrl1 *const timer, const Timer_InterruptHandler handler)
+{
+#ifdef MOCK_REGISTERS
+    baseApbctrl1Registers = malloc(sizeof(Timer_Apbctrl1_Base_Registers));
```

Comment:
If we init more than one timer, we may have a memory leak. If we are using mock registers, and we define the global variable anyway, why malloc at all?

Reply:
When MOCK_REGISTERS defined we need to alloc memory to the baseApbctrl1Registers/baseApbctrl2Registers because those are pointers. Memory leakage protection added.